### PR TITLE
More ray differentials work

### PR DESCRIPTION
**Pull request by est77**
- Added specular reflection and transmission differentials.
- Use specular diffs + an heuristic for other BSDFs.
- Added a ray spread diagnostic surface shader.

### CODE REVIEW

File: src/appleseed/renderer/modeling/bsdf/bsdfsample.cpp [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+
+// Interface header.
+#include "bsdfsample.h"
+
+// appleseed.renderer headers.
+#include "renderer/modeling/bsdf/bsdf.h"
+
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/bsdf/bsdfsample.cpp]: You need to include `<algorithm>` and `<cmath>`.

File: src/appleseed/renderer/modeling/surfaceshader/diagnosticsurfaceshader.cpp [hunk 2]
```diff
@@ -461,6 +463,67 @@ void DiagnosticSurfaceShader::evaluate(
...
+                const BSDF* bsdf = material->get_bsdf();
+                if (bsdf)
+                {
+                    InputEvaluator input_evaluator(shading_context.get_texture_cache());
+                    bsdf->evaluate_inputs(
+                        shading_context,
+                        input_evaluator,
+                        shading_point);
+
+                    Dual3d outgoing(
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/surfaceshader/diagnosticsurfaceshader.cpp]: Make const.

File: src/appleseed/renderer/modeling/surfaceshader/diagnosticsurfaceshader.cpp [hunk 3]
```diff
@@ -461,6 +463,67 @@ void DiagnosticSurfaceShader::evaluate(
...
+                        false,
+                        sample);
+
+                    if (!sample.get_incoming().has_derivatives())
+                        return;
+
+                    // The 3.0 factor is choosen so that ray spread
+                    // from lambertian BRDFs is approx. 1.
+                    const double spread =
+                        std::max(
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/surfaceshader/diagnosticsurfaceshader.cpp]: Remove `std` qualifier, and include `<algorithm>`.

**Comment by est77** [comment]
Fixed and rebased.

**Pull request merged**